services/messaging/src/test/java/org/eclipse/hono/messaging/ForwardingDownstreamAdapterTest.java

```diff
@@ -265,6 +266,47 @@ public void testDownstreamDisconnectClosesUpstreamReceivers() {
         assertTrue(adapter.isSendersPerConnectionEmpty());
     }
 
+    /**
+     * Verifies that the sender and all links to upstream clients are closed when the downstream link is lost
+     */
+    @SuppressWarnings("unchecked")
+    @Test
+    public void testDownstreamLinkClosesUpstreamReceivers() {
+
+        final UpstreamReceiver client = newClient();
+        final ProtonSender downstreamSender = newMockSender(false);
+
+        ArgumentCaptor<Handler<AsyncResult<ProtonSender>>> closeHandlerCaptor = ArgumentCaptor.forClass((Class) Handler.class);
+
+        when(downstreamSender.close()).then(invocation -> {
+            closeHandlerCaptor.getValue().handle(Future.succeededFuture(downstreamSender));
+            return downstreamSender;
+        });
+        when(downstreamSender.closeHandler(closeHandlerCaptor.capture())).then(invocation -> {
+            return downstreamSender;
+        });
+
```

Comment:
IMHO the test should be based on something similar to `HandlerCapturingConnectionFactory` but for a sender, e.g. a `HandlerCapturingProtonSender` which offers a method to get at the registered close or detach handlers.

The whole test case would then be much simpler, I guess. You would just `adapter.addSender(client, downstreamSender)` and then trigger the detach handler of the sender and verify that the upstream client has been closed, right?